src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -152,27 +157,178 @@ public void AddPackageReference(string projectPath, LibraryDependency libraryDep
+            var item = itemGroup.AddItem(PACKAGE_VERSION_TYPE_TAG, libraryDependency.Name);
+            var packageVersion = AddVersionMetadata(libraryDependency, item);
+
+            Logger.LogInformation(string.Format(CultureInfo.CurrentCulture,
+            Strings.Info_AddPkgAdded,
+            libraryDependency.Name,
+            packageVersion,
+            itemGroup.ContainingProject.FullPath
+            ));
+
+            AddExtraMetadataToProjectItemElement(libraryDependency, item);
+        }
+
+        /// <summary>
+        /// Add package name and version into item group.
+        /// </summary>
+        /// <param name="itemGroup">Item group to add to.</param>
+        /// <param name="libraryDependency">Package Dependency of the package to be added.</param>
+        private void AddPackageReferenceIntoItemGroup(ProjectItemGroupElement itemGroup,
+            LibraryDependency libraryDependency)
+        {
+            // Add both package reference information and version metadata using the PACKAGE_REFERENCE_TYPE_TAG.
+            var item = itemGroup.AddItem(PACKAGE_REFERENCE_TYPE_TAG, libraryDependency.Name);
+            var packageVersion = AddVersionMetadata(libraryDependency, item);
+
+            Logger.LogInformation(string.Format(CultureInfo.CurrentCulture, Strings.Info_AddPkgAdded, libraryDependency.Name, packageVersion, itemGroup.ContainingProject.FullPath));
+
+            AddExtraMetadataToProjectItemElement(libraryDependency, item);
+        }
+
+        /// <summary>
+        /// Add only the package name into the project file for projects onboarded to CPM.
+        /// </summary>
+        /// <param name="project">Project to be modified.</param>
+        /// <param name="itemGroup">Item group to add to.</param>
+        /// <param name="libraryDependency">Package Dependency of the package to be added.</param>
+        private void AddPackageReferenceIntoItemGroupCPM(Project project, ProjectItemGroupElement itemGroup,
+            LibraryDependency libraryDependency)
+        {
+            // Only add the package reference information using the PACKAGE_REFERENCE_TYPE_TAG.
+            ProjectItemElement item = itemGroup.AddItem(PACKAGE_REFERENCE_TYPE_TAG, libraryDependency.Name);
+
+            Logger.LogInformation(string.Format(CultureInfo.CurrentCulture, Strings.Info_AddPkgCPM, libraryDependency.Name, project.GetPropertyValue(DirectoryPackagesPropsPathPropertyName), itemGroup.ContainingProject.FullPath));
+
+            AddExtraMetadataToProjectItemElement(libraryDependency, item);
+        }
+
+        /// <summary>
+        /// Add other metadata based on certain flags.
+        /// </summary>
+        /// <param name="libraryDependency">Package Dependency of the package to be added.</param>
+        /// <param name="item">Item to add the metadata to.</param>
+        private void AddExtraMetadataToProjectItemElement(LibraryDependency libraryDependency, ProjectItemElement item)
```

Comment:
I think you have a bunch of great utility methods here. 
It'd be great to get some unit test. There's a lot of logic here that can probably be modified and no test would catch it.

Reply:
That is a good point. Originally, I did consider adding unit tests however a lot of these utility methods were actually just created using code that already existed before I made modifications for CPM projects (I just put them into smaller methods)/they use the MSBuild API which is already tested. Considering that, I opted to write functional tests. 
I am totally open to the suggestion of unit tests though, so please let me know!

Reply:
I think adding better coverage as unit tests is always ok. 

When you separate out a method even if it's from code that's already tested, it is something that can technically be misused. 

You also don't have to add tests for every little method, rather ones where it's convenient.